File: src/engine/smooth.ts
/** Guided filter (He et al., 2010) over Lab planes — SPEC v3 Task 5.5.
 * Edge-preserving smoothing: kills photo grain/texture WITHOUT blurring
 * region boundaries. Implemented with O(N) box filters (moving sums),
 * no integral-image allocations beyond a couple of Float64 planes.
 *
 * Self-guided variant: the image guides itself per channel, which is the
 * standard fast approximation for denoising and is plenty for regionizing.
 */

/** Box (moving-average) filter of a single-channel plane, radius r, O(N). */
function boxFilter(src: Float64Array, w: number, h: number, r: number): Float64Array {
  const tmp = new Float64Array(src.length)
  const out = new Float64Array(src.length)

  // Horizontal pass: sliding window sum.
  for (let y = 0; y < h; y++) {
    const row = y * w
    let acc = 0
    for (let x = -r; x <= r; x++) acc += src[row + Math.min(w - 1, Math.max(0, x))]
    const denom = 2 * r + 1
    for (let x = 0; x < w; x++) {
      tmp[row + x] = acc / denom
      const add = src[row + Math.min(w - 1, x + r + 1)]
      const sub = src[row + Math.max(0, x - r)]
      acc += add - sub
    }
  }

  // Vertical pass: sliding window sum over the horizontal result.
  const denom = 2 * r + 1
  for (let x = 0; x < w; x++) {
    let acc = 0
    for (let y = -r; y <= r; y++) acc += tmp[Math.min(h - 1, Math.max(0, y)) * w + x]
    for (let y = 0; y < h; y++) {
      out[y * w + x] = acc / denom
      const add = tmp[Math.min(h - 1, y + r + 1) * w + x]
      const sub = tmp[Math.max(0, y - r) * w + x]
      acc += add - sub
    }
  }
  return out
}

function mul(a: Float64Array, b: Float64Array): Float64Array {
  const out = new Float64Array(a.length)
  for (let i = 0; i < a.length; i++) out[i] = a[i] * b[i]
  return out
}

function addScaled(a: Float64Array, b: Float64Array, s: Float64Array): Float64Array {
  // Guided-filter output: q = ā·I + b̄
  const out = new Float64Array(a.length)
  for (let i = 0; i < a.length; i++) out[i] = a[i] * b[i] + s[i]
  return out
}

/** Self-guided filter of one plane. eps controls edge preservation:
 * small eps → sharper edges preserved, more texture kept; larger → smoother. */
function guidedPlane(p: Float64Array, w: number, h: number, r: number, eps: number): Float64Array {
  const meanI = boxFilter(p, w, h, r)
  const meanII = boxFilter(mul(p, p), w, h, r)
  // var = E[I²] − E[I]² ; a = var/(var+eps), b = mean − a·mean
  const varI = new Float64Array(p.length)
  const a = new Float64Array(p.length)
  const b = new Float64Array(p.length)
  for (let i = 0; i < p.length; i++) {
    varI[i] = Math.max(0, meanII[i] - meanI[i] * meanI[i])
    a[i] = varI[i] / (varI[i] + eps)
    b[i] = meanI[i] * (1 - a[i])
  }
  const meanA = boxFilter(a, w, h, r)
  const meanB = boxFilter(b, w, h, r)
  return addScaled(meanA, p, meanB)
}

export interface SmoothOptions {
  /** Box radius in pixels (default 4 — ≈ texture scale at working res). */
  radius?: number
  /** Regularization (default 300 in Lab units² — strong edge preservation). */
  eps?: number
}

/** Smooth all three Lab planes with a guided filter. Returns a new buffer. */
export function guidedSmoothLab(
  lab: Float64Array,
  w: number,
  h: number,
  opts: SmoothOptions = {},
): Float64Array {
  const r = opts.radius ?? 4
  const eps = opts.eps ?? 300
  const n = w * h
  const out = new Float64Array(n * 3)
  for (let c = 0; c < 3; c++) {
    const plane = new Float64Array(n)
    for (let i = 0; i < n; i++) plane[i] = lab[i * 3 + c]
    const smooth = guidedPlane(plane, w, h, r, eps)
    for (let i = 0; i < n; i++) out[i * 3 + c] = smooth[i]
  }
  return out
}

File: ios/PaintEngine/Sources/PaintEngine/Smooth.swift
/// Guided filter (He et al., 2010) over Lab planes — port of src/engine/smooth.ts.
/// Edge-preserving smoothing: kills photo grain/texture WITHOUT blurring
/// region boundaries. O(N) box filters (moving sums).
///
/// v3.1 defaults are SOFTENED (r=3, eps=80): the old r=4/eps=300 flattened
/// fine structures (lace, embroidery, blush) before segmentation saw them.

/// Box (moving-average) filter of a single-channel plane, radius r, O(N).
func boxFilter(_ src: [Double], _ w: Int, _ h: Int, _ r: Int) -> [Double] {
    var tmp = [Double](repeating: 0, count: src.count)
    var out = [Double](repeating: 0, count: src.count)
    let denom = Double(2 * r + 1)

    for y in 0..<h {
        let row = y * w
        var acc = 0.0
        for x in -r...r { acc += src[row + min(w - 1, max(0, x))] }
        for x in 0..<w {
            tmp[row + x] = acc / denom
            let add = src[row + min(w - 1, x + r + 1)]
            let sub = src[row + max(0, x - r)]
            acc += add - sub
        }
    }

    for x in 0..<w {
        var acc = 0.0
        for y in -r...r { acc += tmp[min(h - 1, max(0, y)) * w + x] }
        for y in 0..<h {
            out[y * w + x] = acc / denom
            let add = tmp[min(h - 1, y + r + 1) * w + x]
            let sub = tmp[max(0, y - r) * w + x]
            acc += add - sub
        }
    }
    return out
}

/// Self-guided filter of one plane. eps controls edge preservation.
func guidedPlane(_ p: [Double], _ w: Int, _ h: Int, _ r: Int, _ eps: Double) -> [Double] {
    let meanI = boxFilter(p, w, h, r)
    let p2 = p.map { $0 * $0 }
    let meanII = boxFilter(p2, w, h, r)
    var a = [Double](repeating: 0, count: p.count)
    var b = [Double](repeating: 0, count: p.count)
    for i in 0..<p.count {
        let varI = max(0, meanII[i] - meanI[i] * meanI[i])
        a[i] = varI / (varI + eps)
        b[i] = meanI[i] * (1 - a[i])
    }
    let meanA = boxFilter(a, w, h, r)
    let meanB = boxFilter(b, w, h, r)
    var out = [Double](repeating: 0, count: p.count)
    for i in 0..<p.count { out[i] = meanA[i] * p[i] + meanB[i] }
    return out
}

public struct SmoothOptions {
    /// Box radius in pixels (v3.1 default 3).
    public var radius: Int
    /// Regularization in Lab units² (v3.1 default 80 — softer than before).
    public var eps: Double

    public init(radius: Int = 3, eps: Double = 80) {
        self.radius = radius
        self.eps = eps
    }
}

/// Smooth all three Lab planes with a guided filter. Returns a new buffer.
public func guidedSmoothLab(_ lab: [Double], _ w: Int, _ h: Int, opts: SmoothOptions = SmoothOptions()) -> [Double] {
    let n = w * h
    var out = [Double](repeating: 0, count: n * 3)
    for c in 0..<3 {
        var plane = [Double](repeating: 0, count: n)
        for i in 0..<n { plane[i] = lab[i * 3 + c] }
        let smooth = guidedPlane(plane, w, h, opts.radius, opts.eps)
        for i in 0..<n { out[i * 3 + c] = smooth[i] }
    }
    return out
}
